x/batch/types/msg_batch.go

```diff
@@ -0,0 +1,85 @@
+package types
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/axelarnetwork/utils/slices"
+)
+
+var _ types.UnpackInterfacesMessage = BatchRequest{}
+
+// NewBatchRequest is the constructor for BatchRequest
+func NewBatchRequest(sender sdk.AccAddress, messages []sdk.Msg) *BatchRequest {
+	f := func(msg sdk.Msg) types.Any {
+		messageAny, err := cdctypes.NewAnyWithValue(msg)
+		if err != nil {
+			panic(err)
+		}
+		return *messageAny
+	}
+
+	return &BatchRequest{
+		Sender:   sender,
+		Messages: slices.Map(messages, f),
+	}
+}
+
+// Route returns the route for this message
+func (m BatchRequest) Route() string {
+	return RouterKey
+}
+
+// Type returns the type of the message
+func (m BatchRequest) Type() string {
+	return "Batch"
+}
+
+// ValidateBasic executes a stateless message validation
+func (m BatchRequest) ValidateBasic() error {
+	if err := sdk.VerifyAddressFormat(m.Sender); err != nil {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, sdkerrors.Wrap(err, "sender").Error())
+	}
+
+	if len(m.Messages) == 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "empty batch")
+	}
+
+	for _, msg := range m.UnwrapMessages() {
+		if !msg.GetSigners()[0].Equals(m.Sender) {
+			return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "message signer mismatch")
+		}
+
+		if err := msg.ValidateBasic(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GetSigners returns the set of signers for this message
+func (m BatchRequest) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{m.Sender}
+}
+
+// UnpackInterfaces implements UnpackInterfacesMessage
+func (m BatchRequest) UnpackInterfaces(unpacker cdctypes.AnyUnpacker) error {
+	for _, msg := range m.Messages {
+		var sdkMsg sdk.Msg
+		if err := unpacker.UnpackAny(&msg, &sdkMsg); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```

Comment:
Should also call `.UnpackInterfaces` if any of the inner messages implements `UnpackInterfacesMessage`. For example, one of those messages can be a `BatchRequest` itself.

Reply:
nested batch is prevented in ante handler. I prefer batch just unpack as sdk msg, route it, and let its handler to process.

Reply:
I do think the batch message itself should make sure it can't be nested